kernel/src/actions/mod.rs

```diff
@@ -467,9 +474,59 @@ pub(crate) struct CommitInfo {
     /// write this field, but it is optional since many tables will not have this field (i.e. any
     /// tables not written by kernel).
     pub(crate) kernel_version: Option<String>,
-    /// A place for the engine to store additional metadata associated with this commit encoded as
-    /// a map of strings.
-    pub(crate) engine_commit_info: Option<HashMap<String, String>>,
+    /// A place for the engine to store additional metadata associated with this commit
+    pub(crate) engine_info: Option<String>,
+}
+
+impl CommitInfo {
+    pub(crate) fn new(
+        timestamp: i64,
+        operation: Option<String>,
+        engine_info: Option<String>,
+    ) -> Self {
+        Self {
+            timestamp: Some(timestamp),
+            in_commit_timestamp: None,
+            operation: Some(operation.unwrap_or(UNKNOWN_OPERATION.to_string())),
```

Comment:
you're allocating a string to pass to `unwrap_or` each time, even if the `operation` is `Some`. So do this instead:
```suggestion
            operation: Some(operation.unwrap_or_else(||UNKNOWN_OPERATION.to_string())),
```